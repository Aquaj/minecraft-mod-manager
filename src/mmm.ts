#!/usr/bin/env node
import 'dotenv/config';
import { Command } from 'commander';
import { add } from './actions/add.js';
import { list } from './actions/list.js';
import { Platform, ReleaseType } from './lib/modlist.types.js';
import { install } from './actions/install.js';
import { version } from './version.js';
import { update } from './actions/update.js';
import { initializeConfig } from './interactions/initializeConfig.js';
import { Logger } from './lib/Logger.js';

export const APP_NAME = 'Minecraft Mod Manager';
export const APP_DESCRIPTION = 'Manages mods from Modrinth and Curseforge';
export const DEFAULT_CONFIG_LOCATION = './modlist.json';
const cwd = process.cwd();

export interface DefaultOptions {
  config: string;
  debug?: boolean;
  quiet?: boolean;
}

export const program = new Command();

export const logger: Logger = new Logger(program);

export const stop = (): never => {
  // eslint-disable-next-line no-process-exit
  process.exit(-1);
};

const commands = [];

program.name(APP_NAME).version(version).description(APP_DESCRIPTION);

program.on('option:quiet', () => {
  logger.flagQuiet();
});

program.on('option:debug', () => {
  logger.flagDebug();
});

commands.push(
  program.command('list')
    .action(async (_options, cmd) => {
      await list(cmd.optsWithGlobals(), logger);
    })
    .aliases(['l', 'ls'])
);

commands.push(
  program.command('install')
    .action(async (_options, cmd) => {
      await install(cmd.optsWithGlobals(), logger);
    })
    .aliases(['i'])
);

commands.push(
  program.command('update')
    .action(async (_options, cmd) => {
      await update(cmd.optsWithGlobals(), logger);
    })
    .aliases(['u'])
);

commands.push(
  program.command('add')
    .argument('<type>', 'curseforge or modrinth')
    .argument('<id>', 'Curseforge or Modrinth Project Id')
    .action(async (type: Platform, id: string, _options, cmd) => {
      await add(type, id, cmd.optsWithGlobals(), logger);
    })
    .aliases(['a'])
);

commands.push(
  program.command('init')
    .option('-l, --loader <loader>', `Which loader would you like to use? ${Object.values(Platform).join(', ')}`)
    .option('-g, --game-version <gameVersion>', 'What exact Minecraft version are you using? (eg: 1.18.2, 1.19, 1.19.1)')
    .option('-f, --allow-version-fallback', 'Should we try to download mods for previous Minecraft versions if they do not exists for your Minecraft Version?')
    .option('-r, --default-allowed-release-types <defaultAllowedReleaseTypes>',
      `Which types of releases would you like to consider to download? ${Object.values(ReleaseType).join(', ')} - comma separated list`)
    .option('-m, --mods-folder <modsFolder>', `where is your mods folder? (full or relative path from ${cwd})`)
    .action(async (_options, cmd) => {
      await initializeConfig(cmd.optsWithGlobals(), cwd);
    })
);

program.option('-c, --config <MODLIST_JSON>', 'An alternative JSON file containing the configuration', DEFAULT_CONFIG_LOCATION);
program.option('-q, --quiet', 'Suppress all output', false);
program.option('-d, --debug', 'Enable debug messages', false);

