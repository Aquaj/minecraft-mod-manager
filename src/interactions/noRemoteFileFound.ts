import { Logger } from '../lib/Logger.js';
import { DefaultOptions } from '../mmm.js';
import chalk from 'chalk';
import { ModsJson, Platform } from '../lib/modlist.types.js';
import inquirer from 'inquirer';

interface NoRemoteFileFoundInteractionResult {
  platform: Platform;
  id: string;
}

export const noRemoteFileFound = async (
  modName: string,
  platform: Platform,
  configuration: ModsJson,
  logger: Logger,
  options: DefaultOptions
): Promise<NoRemoteFileFoundInteractionResult> => {
  const errorText = `Could not find a file for ${modName} and the Minecraft version ${chalk.whiteBright(configuration.gameVersion)} `
    + `for ${chalk.whiteBright(configuration.loader)} loader`;

  if (options.quiet === true) {
    logger.error(errorText);
    return {} as NoRemoteFileFoundInteractionResult; // needs a return for testing purposes because the above line terminates the process in production
  }

  const newPlatform = platform === Platform.CURSEFORGE ? Platform.MODRINTH : Platform.CURSEFORGE;

  const answers = await inquirer.prompt([
    {
      name: 'confirm',
      type: 'confirm',
      message: `${errorText}. Would you like to try on ${newPlatform}?`,
      default: true
    }
  ]);

  if (answers.confirm === false) {
    logger.error('Aborting', 0);
    return {} as NoRemoteFileFoundInteractionResult; // needs a return for testing purposes because the above line terminates the process in production
  }

  const { newModName } = await inquirer.prompt([
    {
      type: 'input',
      name: 'newModName',
      message: `What is the project id of the mod you want to add on ${newPlatform}?`
    }
  ]);

  return {
    id: newModName,
    platform: newPlatform
  };

};
