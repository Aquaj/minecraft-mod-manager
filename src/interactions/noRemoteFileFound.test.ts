import { afterEach, describe, expect, it, vi } from 'vitest';
import { Platform } from '../lib/modlist.types.js';
import { logger } from '../mmm.js';
import inquirer from 'inquirer';
import { chance } from 'jest-chance';
import { generateModsJson } from '../../test/modlistGenerator.js';
import { noRemoteFileFound } from './noRemoteFileFound.js';

vi.mock('../mmm.js');
vi.mock('inquirer');

describe('The mod not found interaction', () => {
  afterEach(() => {
    vi.resetAllMocks();
  });

  it('shows the correct error message when in quiet mode', async () => {
    const testPlatform = Platform.CURSEFORGE;
    const testModId = 'test-mod-id';
    const randomConfig = generateModsJson().generated;

    await noRemoteFileFound(testModId, testPlatform, randomConfig, logger, { config: 'config.json', quiet: true });

    const loggerErrorCall = vi.mocked(logger.error).mock.calls[0][0];

    expect(loggerErrorCall).toMatchInlineSnapshot('"Could not find a file for test-mod-id and the Minecraft version cuodja for forge loader"');
    expect(vi.mocked(inquirer.prompt)).not.toHaveBeenCalled();

  });

  it('aborts when the user does not want to modify their search', async () => {
    const testPlatform = chance.pickone(Object.values(Platform));
    const testModId = chance.word();
    const randomConfig = generateModsJson().generated;

    vi.mocked(inquirer.prompt).mockResolvedValueOnce({ confirm: false });

    await noRemoteFileFound(testModId, testPlatform, randomConfig, logger, { config: 'config.json', quiet: false });

    const loggerErrorCall = vi.mocked(logger.error).mock.calls[0][0];

    expect(loggerErrorCall).toMatchInlineSnapshot('"Aborting"');
    expect(vi.mocked(inquirer.prompt)).toHaveBeenCalledTimes(1);

  });

  describe.each([
    { input: Platform.CURSEFORGE, expected: Platform.MODRINTH },
    { input: Platform.MODRINTH, expected: Platform.CURSEFORGE }
  ])('when the user wants to modify their search', ({ input, expected }) => {
    it(`it asks for ${expected} when they come from ${input}`, async () => {
      const testPlatform = input;
      const testModId = chance.word();
      const randomConfig = generateModsJson().generated;

      vi.mocked(inquirer.prompt).mockResolvedValueOnce({ confirm: true });
      vi.mocked(inquirer.prompt).mockResolvedValueOnce({ newModName: 'new-mod-id' });

      const actual = await noRemoteFileFound(testModId, testPlatform, randomConfig, logger, {
        config: 'config.json',
        quiet: false
      });

      const inquirerPromptCallArgs = vi.mocked(inquirer.prompt).mock.calls[1][0];

      expect(actual).toEqual({ id: 'new-mod-id', platform: expected });

      expect(inquirerPromptCallArgs).toMatchSnapshot();

    });
  });
});
