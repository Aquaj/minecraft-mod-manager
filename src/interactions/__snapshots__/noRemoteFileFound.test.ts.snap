// Vitest Snapshot v1

exports[`The mod not found interaction > when the user wants to modify their search > it asks for curseforge when they come from modrinth 1`] = `
[
  {
    "message": "What is the project id of the mod you want to add on curseforge?",
    "name": "newModName",
    "type": "input",
  },
]
`;

exports[`The mod not found interaction > when the user wants to modify their search > it asks for modrinth when they come from curseforge 1`] = `
[
  {
    "message": "What is the project id of the mod you want to add on modrinth?",
    "name": "newModName",
    "type": "input",
  },
]
`;
